Wandb callback

manangoel99 opened the pull request:
### PR types
New features

### PR changes
Others

### Describe
This PR fixes #46647 by adding a new `WandbCallback` which automatically logs training, validation and systems metrics to a wandb dashboard like the one shown here.   

<img width="1368" alt="Screenshot 2022-10-11 at 10 42 47 PM" src="https://user-images.githubusercontent.com/32260633/195156712-87de5b09-3951-4add-b06a-a53239ba3fdb.png">
<img width="1376" alt="Screenshot 2022-10-11 at 10 43 06 PM" src="https://user-images.githubusercontent.com/32260633/195156778-dcd0c10f-fa6c-4fdd-ab1a-29082c808009.png">

It also works out of the box for distributed training. Script to reproduce these plots is given in the comments.

Another proposed change is adding the option to add a custom error message to the `paddle.utils.try_import` function.

manangoel99 committed:
- Updated lazy_import to use custom warning messages
- Created new WandbCallback to log training and evaluation metrics to
wandb

manangoel99 committed:
Merge branch 'PaddlePaddle:develop' into WandbCallback

manangoel99 committed:
Merge branch 'PaddlePaddle:develop' into WandbCallback

manangoel99 committed:
Fixed bug during distributed training

manangoel99 commented:
```
import paddle
import paddle.vision.transforms as T
from paddle.static import InputSpec

inputs = [InputSpec([-1, 1, 28, 28], 'float32', 'image')]
labels = [InputSpec([None, 1], 'int64', 'label')]

transform = T.Compose([
    T.Transpose(),
    T.Normalize([127.5], [127.5])
])

train_dataset = paddle.vision.datasets.MNIST(mode='train', transform=transform)
eval_dataset = paddle.vision.datasets.MNIST(mode='test', transform=transform)

net = paddle.vision.models.LeNet()
model = paddle.Model(net, inputs, labels)

optim = paddle.optimizer.Adam(0.001, parameters=net.parameters())
model.prepare(optimizer=optim,
            loss=paddle.nn.CrossEntropyLoss(),
            metrics=paddle.metric.Accuracy())

callback = paddle.callbacks.WandbCallback(project='paddle_mnist')
model_callback = paddle.callbacks.ModelCheckpoint('./checkpoint')
model.fit(train_dataset, eval_dataset, batch_size=64, callbacks=[callback, model_callback], epochs=5)

```

manangoel99 commented:
@heavengate @jzhang533 It would be awesome if I could get a review on this PR

manangoel99 committed:
Minor typo fix

manangoel99 committed:
Fixed flake8 errors

manangoel99 committed:
Fixed code formatting errors

AndPuQing commented:
You may need to add a unit test, like https://github.com/PaddlePaddle/Paddle/blob/develop/python/paddle/tests/test_callback_visualdl.py

jzhang533 commented on python/paddle/hapi/callbacks.py:
```diff
@@ -957,6 +957,175 @@ def on_eval_end(self, logs=None):
+        entity(str, optional): Name of the team/user creating the run. Default: Logged in user
+        name(str, optional): Name of the run. Default: randomly generated by wandb
+        dir(str, optional): Directory in which all the metadata is stored. Default: `wandb`
+        offline(str, optional): Can be "online", "offline" or "disabled". Default: "online".
```
this should be `mode` arg?

manangoel99 replied:
Oh yes! Fixing it now!

jzhang533 reviewed (commented):
is it possible to have a testcase to cover this feature?

manangoel99 commented:
For implementing a unit test, the wandb python library would have to be added to the requirements.txt file for the package. Is that an acceptable change?

AndPuQing commented:
> For implementing a unit test, the wandb python library would have to be added to the requirements.txt file for the package. Is that an acceptable change?

add it in https://github.com/PaddlePaddle/Paddle/blob/develop/python/unittest_py/requirements.txt . just import it when utest.

manangoel99 committed:
Added unit test for wandb callback

manangoel99 committed:
Fixed code formatting errors

manangoel99 committed:
unit tests requirements file to add wandb

manangoel99 committed:
unit tests requirements file to add wandb

manangoel99 committed:
Merge branch 'WandbCallback' of https://github.com/manangoel99/Paddle into WandbCallback

manangoel99 committed:
Merge branch 'PaddlePaddle:develop' into WandbCallback

manangoel99 committed:
Fixed formatting for WandbCallback unit tests

manangoel99 committed:
Merge branch 'WandbCallback' of https://github.com/manangoel99/Paddle into WandbCallback

heavengate commented on python/paddle/hapi/callbacks.py:
```diff
@@ -957,6 +957,175 @@ def on_eval_end(self, logs=None):
+        self._run = None
+        self.wandb_args.update(**kwargs)
+
+        _ = self.run
```
What does this line do? `_` is usually used as unused variable.

manangoel99 replied:
This line initializes the wandb run if a run is not already in progess in which case it uses the ongoing run by calling this method

https://github.com/manangoel99/Paddle/blob/8353b6996d4bc157a6dd416ecbadb35df3c75346/python/paddle/hapi/callbacks.py#L1039-L1053

heavengate commented on python/paddle/tests/test_callback_wandb.py:
```diff
@@ -0,0 +1,71 @@
+# Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
```
2020 -> 2022

heavengate commented on python/paddle/tests/test_callback_wandb.py:
```diff
@@ -0,0 +1,71 @@
+        return 512
+
+
+class TestCallbacks(unittest.TestCase):
```
TestCallbacks -> TestWandbCallbacks

manangoel99 committed:
Fixed errors in WandbCallback unit test

heavengate commented:
There are some errors in CI tests， you can click Details for logs:
<img width="842" alt="截屏2022-10-12 下午5 08 24" src="https://user-images.githubusercontent.com/12605721/195300999-53e41c8c-711c-4c9b-bfc0-59cf3aee81c9.png">
1. Static Check: CI with extract and run examples code from Python comments,  it seems paddle is not imported in exmaple codes
<img width="863" alt="截屏2022-10-12 下午5 11 14" src="https://user-images.githubusercontent.com/12605721/195301694-62d9c00c-59cb-4066-b2ac-9638adb9d5b8.png">
2.  test_callback_wandb run failed or timeout in Mac/Windows, please check locally
<img width="873" alt="截屏2022-10-12 下午5 13 12" src="https://user-images.githubusercontent.com/12605721/195302168-df6cc9bf-b174-4370-8866-a02268ce5d58.png">

manangoel99 commented:
@heavengate Thank you for the detailed review! I'll look into these errors and update the PR

manangoel99 committed:
Fix WandbCallback test

manangoel99 committed:
Merge branch 'PaddlePaddle:develop' into WandbCallback

manangoel99 committed:
Fixed error in tests and Fixed code formatting errors

manangoel99 committed:
Fixed error in tests

manangoel99 committed:
Merge branch 'WandbCallback' of https://github.com/manangoel99/Paddle into WandbCallback

manangoel99 committed:
Merge branch 'PaddlePaddle:develop' into WandbCallback

manangoel99 commented:
@heavengate for the Mac build I'm getting an import error even though the `wandb` package has been added to the `python/unittest_py/requirements.txt` file. I'm not sure how to fix this.

<img width="855" alt="Screenshot 2022-10-18 at 4 06 29 PM" src="https://user-images.githubusercontent.com/32260633/196407668-57bcef91-7528-48f4-8000-037bb7a1154b.png">

Also for the Static check for the example codes, it is giving an ImportError that `import paddle` has not been called even though the line is [here](https://github.com/manangoel99/Paddle/blob/6311dfcca85461dc0a6495374d6722b0dcc8a3b4/python/paddle/hapi/callbacks.py#L983)

Could you guide me on how to fix these

manangoel99 committed:
Merge branch 'PaddlePaddle:develop' into WandbCallback

heavengate commented:
> @heavengate for the Mac build I'm getting an import error even though the `wandb` package has been added to the `python/unittest_py/requirements.txt` file. I'm not sure how to fix this.
> 
> <img alt="Screenshot 2022-10-18 at 4 06 29 PM" width="855" src="https://user-images.githubusercontent.com/32260633/196407668-57bcef91-7528-48f4-8000-037bb7a1154b.png">
> 
> Also for the Static check for the example codes, it is giving an ImportError that `import paddle` has not been called even though the line is [here](https://github.com/manangoel99/Paddle/blob/6311dfcca85461dc0a6495374d6722b0dcc8a3b4/python/paddle/hapi/callbacks.py#L983)
> 
> Could you guide me on how to fix these

It seems `wandb` install failed due to environment, I rerun the CI and it is already passed

luotao1 commented:
`test_callback_wandb (Timeout)` in PR-CI-Coverage
```
2022-10-20 15:55:18     Test #464: test_callback_wandb ..............***Timeout  15.00 sec
```
How much time of test_callback_wandb? If you need much more time, you could `set_tests_properties` like：  
https://github.com/PaddlePaddle/Paddle/blob/ef80073cb82cf4341ff9ebb547ee2694f49186cd/python/paddle/tests/CMakeLists.txt#L59

manangoel99 committed:
Increased timeout for wandb callback test

manangoel99 commented:
@luotao1 Thank you for the suggestion! I've added that

manangoel99 committed:
Update CMakeLists.txt

luotao1 commented on python/paddle/tests/CMakeLists.txt:
```diff
@@ -66,6 +66,7 @@ set_tests_properties(test_vision_models PROPERTIES TIMEOUT 120)
 set_tests_properties(test_dataset_uci_housing PROPERTIES TIMEOUT 120)
 set_tests_properties(test_dataset_imdb PROPERTIES TIMEOUT 300)
 set_tests_properties(test_pretrained_model PROPERTIES TIMEOUT 600)
+set_tests_properties(test_callback_wandb PROPERTIES TIMEOUT 300)
```
Since the elapsed time of `test_callback_wandb` are less than 30 seconds, thus, `TIMEOUT 300`->`TIMEOUT 60`?
```
2022-10-20 18:46:06     Test #463: test_callback_wandb ..............   Passed   28.08 sec
2022-10-20 18:46:06     Start 463: test_callback_wandb
2022-10-20 18:46:29     Test #463: test_callback_wandb ..............   Passed   23.44 sec
2022-10-20 18:46:29     Start 463: test_callback_wandb
2022-10-20 18:46:53 1/1 Test #463: test_callback_wandb ..............   Passed   23.59 sec
```

manangoel99 replied:
I've made the requested change

luotao1 commented on python/paddle/utils/lazy_import.py:
```diff
@@ -32,9 +32,10 @@ def try_import(module_name):
+                "Failed importing {}. This likely means that some paddle modules "
+                "require additional dependencies that have to be "
+                "manually installed (usually with `pip install {}`). ").format(
+                    module_name, install_name)
```
<img width="1001" alt="image" src="https://user-images.githubusercontent.com/6836917/197105835-bfab9bc2-bb58-4e07-a382-1d5c353ce1f4.png">

<img width="717" alt="image" src="https://user-images.githubusercontent.com/6836917/197105770-d9d3629c-4e18-4fd6-9ea7-9570299f5686.png">
The change here doesn't pass the python coverage test, could you add a unit-test or revert the change (use the default error message)?

manangoel99 replied:
I added aa unit test. Thank you for the suggestion

luotao1 commented on python/unittest_py/requirements.txt:
```diff
@@ -17,3 +17,4 @@ numpy>=1.20,<1.22; python_version >= "3.7"
 autograd==1.4
 librosa==0.8.1
 parameterized
+wandb
```
How about specific the version of wandb?

manangoel99 committed:
Merge branch 'PaddlePaddle:develop' into WandbCallback

manangoel99 committed:
- Changed timeout for wandb callback test
- Added test for testing custom error message in try_import
- Added version for wandb in unittest requirements

manangoel99 committed:
Fixed code formatting in lazy import test

luotao1 commented:
LGTM, please solve the code conflicts

manangoel99 committed:
Fix merge conflicts with develop

manangoel99 commented:
@luotao1 I've fixed the merge conflicts and updated the PR

manangoel99 committed:
Fix code formatting errors

luotao1 reviewed (approved):
LGTM

jzhang533 reviewed (approved):
LGTM

XieYunshen reviewed (approved):
LGTM

jeff41404 reviewed (approved):
LGTM

luotao1 commented:
@manangoel99 Thanks very much for your contribution!
- We need a Chinese API document for `paddle.callbacks.WandbCallback`, like [`paddle.callbacks`](https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/api/paddle/callbacks/Overview_cn.html). The source code of the Chinese API document should be put in  [docs#api/paddle/callbacks](https://github.com/PaddlePaddle/docs/tree/develop/docs/api/paddle/callbacks). Since we find https://docs.wandb.ai/v/zh-hans/ have a Chinese version, could you introduce a developer to contribute the Chinese API document for `paddle.callbacks.WandbCallback`?
- We have a bi-week online Paddle framework contributor meetup, if you have interest to give a talk to introduce your project at this event, please let us know.

manangoel99 commented:
@luotao1 Sure! We can add the Chinese API documentation as well. Can you also tell me where the english documentation would go so that I can write up a more detailed bit of documentation or is it built from the docstring?

I would love to be a part of the contributor meet up and talk to the people about what this PR does and how Weights & Biases is useful

luotao1 commented:
> Can you also tell me where the English documentation would go so that I can write up a more detailed bit of documentation or is it built from the docstring?

Yes, English documentation is built from the docstring, like [`paddle.callbacks.Callback`](https://www.paddlepaddle.org.cn/documentation/docs/en/develop/api/paddle/callbacks/Callback_en.html) in our website. Our website needs some time to build the latest develop branch, we will tell you if `paddle.callbacks.WandbCallback` is shown on the website.

> I would love to be a part of the contributor meet up and talk to the people about what this PR does and how Weights & Biases is useful

Got it. We will discuss the details of the meetup event internally first and then discuss it with you.

luotao1 commented:
@manangoel99 Our website updates, you can see https://www.paddlepaddle.org.cn/documentation/docs/en/develop/api/paddle/callbacks/WandbCallback_en.html

luotao1 commented:
>  We will discuss the details of the meetup event internally first

How about: 7-8 p.m., November 24, Beijing time ?
Tencent Meeting: https://apps.apple.com/cn/app/tencent-meeting/id1484048379?l=en

manangoel99 commented:
@luotao1 sounds good! Thank you!
